### PR TITLE
Fixes environment variable prefixes with double underscores preventing environment variables from being parsed

### LINKED ISSUE #40911: Environment Variable configuration provider doesn't work if the prefix has 2 underscores
### Description
Environment variable loader doesn't load variables if the prefix is set to a string with one or more double underscores.

The problem lies in the incorrect order of operations here: https://github.com/dotnet/runtime/blob/98eecc721d5ad27561fc4463eb521930c2d21a5f/src/libraries/Microsoft.Extensions.Configuration.EnvironmentVariables/src/EnvironmentVariablesConfigurationProvider.cs#L53-L56

The keys are first normalized (that is `__` is replaced with separator; default `:`), *then* they are filtered. The operations should happen the other way around.

To reproduce the issue:
1. Create a new ASP.NET Core application
2. In Startup.cs, in the constructor, add the following code:
  ```cs
  var cfg = new ConfigurationBuilder().AddEnvironmentVariables("TEST__").Build();
  Console.WriteLine(cfg["Test1:Test2"] ?? "<null>");
  ```
3. In project properties, under Debug, add the following environment variable: `TEST__TEST1__TEST2` with a value of `42`.
4. Run the project.

#### Expected result
Console prints `42`.

#### Actual result
Console prints `<null>`.

### Configuration
This affected my application running on ASP.NET Core 3.1.7, running on Windows 10 Pro 2004 x64, as well as Alpine Linux 3.12.0 x64 (running in Docker).

Browsing source of the master branch reveals the buggy code remains, which suggests the bug is present in .NET 5 as well.

### Regression?
Not as far as I can tell.

### Other information
Tried on Core 3.1, the problem appears to be present in .NET 5 as well.

Currently, the problem can be mitigated by replacing all instances of `__` in the prefix with `:` or whatever other separator was configured, if one was configured. Modifying line 1 from repro to `.AddEnvironmentVariables("TEST:")` will cause the program to print `42` instead of `<null>`.

This caveat is not documented, and therefore I consider it a bug.

The issue might not be trivial to fix, as I've noticed that several other prefixes are tested as well.

In my code, I decided to use `__` for consistency (so that `Test1:Test2` becomes `TEST__TEST1__TEST2` rather than `TEST_TEST1__TEST2`).

### PR DESCRIPTION
**Pull request by Emzi0767**
This PR fixes #40911.

The following changes are implemented:
- `AzureEnvToAppEnv` now filters out variables which do not start with any of the valid prefixes.
- Environment variable name processing is altered as such:
   - The specified variable prefix is stripped from the variable name before normalization.
   - The variable name is normalized.
   - The prefix is then prepended to the variable name.

Relevant tests have been added to the repository. All passing.

### CODE REVIEW

**Comment by dnfadmin** [comment]
[CLA assistant check](https://cla.dotnetfoundation.org/dotnet/runtime?pullRequest=42932) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.dotnetfoundation.org/dotnet/runtime?pullRequest=42932) before we can accept your contribution.<br/><br/>:x: Emzi0767 [sign now](https://cla.dotnetfoundation.org/dotnet/runtime?pullRequest=42932)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.dotnetfoundation.org/check/dotnet/runtime?pullRequest=42932) it.</sub>

**Comment by Emzi0767** [comment]
I should mention: this PR also supersedes #42932, the author of which appears inactive.

File: src/libraries/Microsoft.Extensions.Configuration.EnvironmentVariables/src/EnvironmentVariablesConfigurationProvider.cs [hunk 1]
```diff
@@ -94,12 +96,21 @@ private static IEnumerable<DictionaryEntry> AzureEnvToAppEnv(DictionaryEntry ent
             {
                 prefix = CustomPrefix;
             }
-            else
+            else if (key.StartsWith(envPrefix, StringComparison.OrdinalIgnoreCase))
             {
-                entry.Key = NormalizeKey(key);
+                // This prevents the prefix from being normalized.
+                // We can also do a fast path branch, I guess? No point in reallocating if the prefix is empty.
+                entry.Key = envPrefix != string.Empty
```
Comment (safern) [inline src/libraries/Microsoft.Extensions.Configuration.EnvironmentVariables/src/EnvironmentVariablesConfigurationProvider.cs:103]: if the key starts with the `envPrefix` when would `envPrefix` be `string.Empty`?

**Review by safern — COMMENTED** [review]
Should we support more than double `_`?

File: src/libraries/Microsoft.Extensions.Configuration.EnvironmentVariables/src/EnvironmentVariablesConfigurationProvider.cs (hunk 1 above)
Comment (Emzi0767) [reply src/libraries/Microsoft.Extensions.Configuration.EnvironmentVariables/src/EnvironmentVariablesConfigurationProvider.cs:103]: In the event user does not specify a prefix, it defaults to `string.Empty`. Obviously, for any given string s such that s is not null, `s.StartsWith(string.Empty, StringComparison.OrdinalIgnoreCase)` is true.

**Comment by Emzi0767** [comment]
> Should we support more than double `_`?

This PR simply prevents normalization of the prefix. Number of underscores doesn't matter. So long as it's part of the prefix, it will not be normalized.

File: src/libraries/Microsoft.Extensions.Configuration.EnvironmentVariables/src/EnvironmentVariablesConfigurationProvider.cs (hunk 1 above)
Comment (Emzi0767) [reply src/libraries/Microsoft.Extensions.Configuration.EnvironmentVariables/src/EnvironmentVariablesConfigurationProvider.cs:103]: See lines 38 and 29 for where that assumption comes from.

**Comment by Emzi0767** [comment]
I should also point out, that the error here occured due to the fact that normalization of variable name happened *before* the user-specified prefix was fixed, which led to situations where if variables were prefixed with a string that contained a double-or-longer underscore substring, it was normalized, and then no longer matched the prefix, which meant the variables weren't loaded. This fix was to simply swap the order of those operations.

**Comment by danmosemsft** [comment]
@maryamariyan what is the next step here? this one seems to have got stuck. 😃

File: src/libraries/Microsoft.Extensions.Configuration.EnvironmentVariables/tests/EnvironmentVariablesTest.cs [hunk 2]
```diff
@@ -153,6 +153,38 @@ public void ReplaceDoubleUnderscoreInEnvironmentVariables()
...
+                {
+                    {"test__prefix__with__double__underscores__data__ConnectionString", "connection"},
+                    {"SQLCONNSTR__db1", "connStr"}
+                };
+            var envConfigSrc = new EnvironmentVariablesConfigurationProvider("test__prefix__with__double__underscores__");
+
+            envConfigSrc.Load(dict);
+
+            Assert.Equal("connection", envConfigSrc.Get("data:ConnectionString"));
+            Assert.Equal("System.Data.SqlClient", envConfigSrc.Get("ConnectionStrings:_db1_ProviderName"));
```
Comment (maryamariyan) [inline src/libraries/Microsoft.Extensions.Configuration.EnvironmentVariables/tests/EnvironmentVariablesTest.cs:169]: Could you please also add a test to cover the example in the comment below:

https://github.com/dotnet/runtime/issues/40911#issuecomment-686805871

File: src/libraries/Microsoft.Extensions.Configuration.EnvironmentVariables/src/EnvironmentVariablesConfigurationProvider.cs (hunk 1 above)
Comment (safern) [reply src/libraries/Microsoft.Extensions.Configuration.EnvironmentVariables/src/EnvironmentVariablesConfigurationProvider.cs:103]: Makes sense. It kind of confused me but now that I see this it makes sense.

File: src/libraries/Microsoft.Extensions.Configuration.EnvironmentVariables/src/EnvironmentVariablesConfigurationProvider.cs [hunk 3]
```diff
@@ -94,12 +96,21 @@ private static IEnumerable<DictionaryEntry> AzureEnvToAppEnv(DictionaryEntry ent
...
+                    ? envPrefix + NormalizeKey(key.Substring(envPrefix.Length))
```
Comment (safern) [inline src/libraries/Microsoft.Extensions.Configuration.EnvironmentVariables/src/EnvironmentVariablesConfigurationProvider.cs:104]: use, `string.concat` instead?

**Comment by safern** [comment]
Overall, looks good, @maryamariyan and I left some comments. Also, merge conflicts need to be addressed.

**Comment by Emzi0767** [comment]
I shall address the changes later this week, possibly today

File: src/libraries/Microsoft.Extensions.Configuration.EnvironmentVariables/tests/EnvironmentVariablesTest.cs (hunk 2 above)
Comment (Emzi0767) [reply src/libraries/Microsoft.Extensions.Configuration.EnvironmentVariables/tests/EnvironmentVariablesTest.cs:169]: I will

**Comment by Emzi0767** [comment]
I am stuck with a problem: I am either misunderstanding something or #44923 altered the behaviour of loading SQL connection strings:

Prior to #44923, adding a connectionstring-prefixed envvar to the collection would yield it back in the configuration with `ConnectionStrings:` prefix. Post 44923 it does not. Is this a regression or am I misunderstanding something?

**Comment by Emzi0767** [comment]
Nevermind above, I have tested above and it seems I misunderstood.

**Comment by Emzi0767** [comment]
I apologize for the slight mess in commits, however git is refusing to recognize that I've merged a change.

**Review by safern — APPROVED** [review]
LGTM

**Pull request merged**